File: src/bridge/game.ts
import { invoke } from '@tauri-apps/api/core';
import type { UnlistenFn } from '@tauri-apps/api/event';
import { listen } from '@tauri-apps/api/event';

export async function getCluster(uuid: string): Promise<Core.ClusterWithManifest> {
	return await invoke<Core.ClusterWithManifest>(
		'plugin:onelauncher|get_cluster',
		{ uuid },
	);
}

export async function getClusterLogs(uuid: string): Promise<string[]> {
	return await invoke<string[]>(
		'plugin:onelauncher|get_cluster_logs',
		{ uuid },
	);
}

export async function getClusterLog(uuid: string, log: string): Promise<string> {
	return await invoke<string>(
		'plugin:onelauncher|get_cluster_log',
		{ uuid, log },
	);
}

export async function getClustersGrouped(): Promise<Map<string, WithIndex<Core.ClusterWithManifest>[]>> {
	const clusters = await getClusters();
	const map = new Map<string, WithIndex<Core.ClusterWithManifest>[]>();

	clusters.forEach((cluster, index) => {
		const groupName = cluster.cluster.group || 'Unnamed';
		const value: WithIndex<Core.ClusterWithManifest> = {
			...cluster,
			index,
		};

		if (map.has(groupName))
			map.get(groupName)!.push(value);
		else
			map.set(groupName, [value]);
	});

	return map;
}

export async function getClusters(): Promise<Core.ClusterWithManifest[]> {
	return await invoke<Core.ClusterWithManifest[]>('plugin:onelauncher|get_clusters');
}

export async function createCluster(cluster: Omit<Core.Cluster, 'id' | 'createdAt'> & { version: string }): Promise<void> {
	return await invoke(
		'plugin:onelauncher|create_cluster',
		{
			name: cluster.name,
			version: cluster.version,
			client: cluster.client,
			// cover: cluster.cover,
			// group: cluster.group,
		},
	);
}

export async function refreshClientManager(): Promise<void> {
	return await invoke('plugin:onelauncher|refresh_client_manager');
}

interface LaunchCallbacks {
	on_launch: (pid: number) => any;
	on_stdout: (line: string) => any;
	on_stderr: (line: string) => any;
};

export function launchCluster(uuid: string, callbacks?: LaunchCallbacks): Promise<number> {
	// eslint-disable-next-line no-async-promise-executor
	return new Promise(async (resolve) => {
		let unlisten_stdout: UnlistenFn | undefined;
		let unlisten_stderr: UnlistenFn | undefined;
		let pid: number = -1;

		function guard(passed_pid: number, fn: () => any | undefined) {
			if (passed_pid !== pid)
				return;

			if (fn)
				fn();
		}

		const unlisten_launch = await listen<number>('game:launch', async (e) => {
			pid = e.payload;

			if (callbacks)
				callbacks.on_launch(pid);

			unlisten_launch!();
		});

		const unlisten_exit = await listen<[number, number]>(`game:exit`, e => guard(e.payload[0], () => {
			if (callbacks) {
				unlisten_stdout!();
				unlisten_stderr!();
			}

			unlisten_exit!();
			resolve(e.payload[1]);
		}));

		if (callbacks) {
			unlisten_stdout = await listen<[number, string]>(`game:stdout`, e => guard(
				e.payload[0],
				callbacks.on_stdout(e.payload[1]),
			));

			unlisten_stderr = await listen<[number, string]>(`game:stderr`, e => guard(
				e.payload[0],
				callbacks.on_stderr(e.payload[1]),
			));
		}

		await invoke<number>(
			'plugin:onelauncher|launch_cluster',
			{ uuid },
		);
	});
}

File: src-tauri/src/api/game.rs
use std::path::PathBuf;

use onelauncher::{game::{clients::LaunchCallbacks, client_manager::ClusterWithManifest, clients::ClientType}, AppState};
use tauri::{Manager, State};
use tokio::sync::Mutex;
use uuid::Uuid;

#[tauri::command]
pub async fn create_cluster(
	state: State<'_, Mutex<AppState>>,
	name: String,
	version: String,
	client: ClientType,
	cover: Option<String>,
	group: Option<Uuid>,
) -> Result<Uuid, String> {
	let manager = &mut state.lock().await.clients;
	let uuid = manager
		.create_cluster(name, version, cover, group, client)
		.await?;
	Ok(uuid)
}

#[tauri::command]
pub async fn get_clusters(
	state: State<'_, Mutex<AppState>>,
) -> Result<Vec<ClusterWithManifest>, String> {
	let manager = &mut state.lock().await.clients;
	Ok(manager.get_clusters_owned())
}

#[tauri::command]
pub async fn get_cluster(
	state: State<'_, Mutex<AppState>>,
	uuid: Uuid,
) -> Result<ClusterWithManifest, String> {
	let manager = &mut state.lock().await.clients;
	let cluster = manager.get_cluster(uuid)?;
	Ok(cluster.clone())
}

#[tauri::command]
pub async fn get_cluster_logs(
    state: State<'_, Mutex<AppState>>,
    uuid: Uuid,
) -> Result<Vec<PathBuf>, String> {
    let manager = &mut state.lock().await.clients;
    let cluster = manager.get_cluster(uuid)?;

    let log_files = cluster.cluster.get_log_files()?;
    Ok(log_files)
}

#[tauri::command]
pub async fn get_cluster_log(
    state: State<'_, Mutex<AppState>>,
    uuid: Uuid,
    log: PathBuf,
) -> Result<String, String> {
    let manager = &mut state.lock().await.clients;
    let cluster = manager.get_cluster(uuid)?;

    let log = cluster.cluster.get_log(&log)?;
    Ok(log)
}

#[tauri::command]
pub async fn launch_cluster(
    app: tauri::AppHandle,
	uuid: Uuid,
) -> Result<(), String> {
    let state = &mut app.state::<Mutex<AppState>>();
    let manager = &mut state.lock().await.clients;

    let on_launch_app = app.clone();
    let on_stdout_app = app.clone();
    let on_stderr_app = app.clone();
    let on_exit_app = app.clone();

    let callbacks = LaunchCallbacks {
        on_launch: Box::new(move |pid| {
            on_launch_app.emit("game:launch", pid).expect("Failed to emit game:launched");
        }),

        on_stdout: Box::new(move |pid, line| {
            on_stdout_app.emit("game:stdout", (pid, line)).expect("Failed to emit game:stdout");
        }),

        on_stderr: Box::new(move |pid, line| {
            on_stderr_app.emit("game:stderr", (pid, line)).expect("Failed to emit game:stderr");
        }),

        on_exit: Box::new(move |pid, exit_code| {
            on_exit_app.emit("game:exit", (pid, exit_code)).expect("Failed to emit game:exit");
        })
    };

    let _ = manager.launch(uuid, callbacks).await?;
    Ok(())
}
